lib/benchmarks/gallery_automator.dart

```diff
@@ -0,0 +1,199 @@
+
+  bool finished = false;
+  LiveWidgetController controller;
+
+  /// An iterable that generates all demo names.
+  Iterable<String> get demoNames =>
+      allGalleryDemos(GalleryLocalizationsEn()).map((demo) => demo.describe);
+
+  Future<void> _handleError(dynamic error, StackTrace stackTrace) async {
+    print('Error: $error');
+    print('StackTrace: $stackTrace');
+    await reportError(error, stackTrace);
+  }
+
+  Widget createWidget() {
+    Future<void>.delayed(
+      _initialWaitingDuration,
+      testScrollsOnly ? automateScrolls : automateDemoGestures,
+    ).catchError(_handleError);
+    return const GalleryApp();
+  }
+
+  /// Opens and quits demos that are specified by [runCriterion], twice.
+  Future<void> automateDemoGestures() async {
+    await animationStops();
+
+    controller = LiveWidgetController(WidgetsBinding.instance);
+
+    reporter('==== List of demos to be run ====');
+    for (final demo in demoNames) {
+      if (!runCriterion(demo)) continue;
+      reporter(demo);
+    }
+    reporter('==== End of list of demos to be run ====');
+
+    var finishedStudyDemos = false;
+
+    for (final demo in demoNames) {
+      if (!finishedStudyDemos && typeOfDemo(demo) != DemoType.study) {
+        finishedStudyDemos = true;
+
+        await scrollUntilVisible(
+          element: find.text('Categories').evaluate().single,
+          strict: true,
+          animated: false,
+        );
+      }
+
+      final demoButton =
+          find.byKey(ValueKey(demo), skipOffstage: false).evaluate().single;
+
+      await scrollUntilVisible(
+        element: demoButton,
+        animated: false,
+        strict: true,
+      );
+
+      // Skip demo if it does not pass `runCriterion`.
+      // This continue statement is placed here because we need to scroll
+      // through each `Scrollable` to find the `demoButton`.
+      if (!runCriterion(demo)) continue;
+
+      reporter('Running demo "$demo"');
+
+      for (var i = 0; i < 2; ++i) {
+        await controller.tap(find.byKey(ValueKey(demo)));
+
+        if (typeOfDemo(demo) == DemoType.animatedWidget) {
+          await Future<void>.delayed(_defaultWaitingDuration);
+        } else {
+          await animationStops();
+        }
+
+        await controller.tap(find.byKey(const ValueKey('Back')));
+
+        await animationStops();
+      }
+    }
+
+    reporter('All demos finished.');
+
+    // At the end of the test, mark as finished.
+    finished = true;
+  }
+
+  /// Scrolls various parts of the gallery.
+  Future<void> automateScrolls() async {
+    reporter('Running scrolling test.');
+    // Extract categories.
+    String categoryOf(String demo) {
+      final atSymbolIndex = demo.lastIndexOf('@');
+      if (atSymbolIndex < 0) {
+        return '';
+      } else {
+        return demo.substring(atSymbolIndex + 1);
+      }
+    }
+
+    // Select a demo from each category, so that we can test scrolling each one.
+    final coveredCategories = <String>{};
```

Comment:
final => final String

Reply:
The type is `Set<String>`.
The Gallery repo also requires omitting local variable types, as shown in this `flutter analyze` report.

```
   info • Omit type annotations for local variables • lib/benchmarks/gallery_automator.dart:156:5 •
          omit_local_variable_types
```